src/realm/array_uuid.hpp

```diff
@@ -0,0 +1,122 @@
+/*************************************************************************
+ *
+ * Copyright 2019 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ **************************************************************************/
+
+#ifndef REALM_ARRAY_UUID_HPP
+#define REALM_ARRAY_UUID_HPP
+
+#include <realm/array.hpp>
+#include <realm/uuid.hpp>
+
+namespace realm {
+
+class ArrayUUID : public ArrayPayload, protected Array {
+public:
+    using value_type = UUID;
+
+    using Array::Array;
+    using Array::destroy;
+    using Array::get_ref;
+    using Array::init_from_mem;
+    using Array::init_from_parent;
+    using Array::update_parent;
+    using Array::verify;
+
+    static UUID default_value(bool nullable)
+    {
+        if (nullable) {
+            return UUID();
+        }
+        // FIXME: verify this and optimize it
+        return UUID("00000000-0000-0000-0000-000000000001");
```

Comment:
Doesn't this seem a bit weird?

Are we totally sure that we should consider the zero-UUID as NULL? Isn't that a perfectly valid UUID?

To me, I would definitely prefer moving away from the (IMO unfortunate) pattern of letting each value type have its own little personal representation of NULL.

Reply:
According to RFC 4122, "The nil UUID is special form of UUID that is specified to have all 128 bits set to zero." I have assumed that we should consider Realm's "null" as the same as this definition. However if there is a good argument to make for having both a Realm null UUID and a nil, it is not too late to change the implementation. Should we distinguish these two definitions?

Reply:
From an SDK perspective, there is a big difference between between returning `null` and an UUID no matter the value.

Reply:
Ok I will change the implementation to allow UUID(000..00) as a valid value, and use `Optional<UUID>{}` as the null. This will be consistent with MongoDB's implementation (thanks to @jbreams for the test!): 

```
rs:PRIMARY> db.bar.find()
{ "_id" : ObjectId("5f5fb26ee10f90a2d9d133cc"), "a" : ObjectId("000000000000000000000000") }
{ "_id" : ObjectId("5f5fb276e10f90a2d9d133cd"), "a" : UUID("00000000-0000-0000-0000-000000000000") }
{ "_id" : ObjectId("5f5fb27ee10f90a2d9d133ce"), "a" : null }
{ "_id" : ObjectId("5f5fb282e10f90a2d9d133cf") }
11:13
rs:PRIMARY> db.bar.find({a: null})
{ "_id" : ObjectId("5f5fb27ee10f90a2d9d133ce"), "a" : null }
{ "_id" : ObjectId("5f5fb282e10f90a2d9d133cf") }
```